### PR TITLE
Add unit tests for Real User Monitoring

### PR DESCRIPTION
**Pull request by rowanmanning**
We're replacing Perfume with Google Web Vitals, and I'm not confident
enough in the way this works to make the required changes. The key thing
to test is that this library emits the expected o-tracking event when it
has all of the RUM data.

Relates-To: [FTDCS-294](https://financialtimes.atlassian.net/browse/FTDCS-294)

### CODE REVIEW

**Review by andygout — APPROVED** [review]
Took me a little while to work through it but all makes sense and I cannot think of any additional scenarios that should be tested.

**Comment by rowanmanning** [comment]
Thanks 🙂  I think there are _maybe_ one or two more complex edge cases we could test, but these are the main ones that will end up changing as a result of the switch to Google Web Vitals

**Review by alexmuller — APPROVED** [review]
This is really useful - it would have taken me much longer to unpick the functionality if I didn't have these tests to work from. Good consistent metric data too.

File: src/client/__test__/realUserMonitoring.spec.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+				});
+
+				it('broadcasts an o-tracking event with the formatted metrics data', () => {
+					expect(broadcast).toHaveBeenCalledTimes(1);
+					const broadcastArguments = broadcast.mock.calls[0];
+					expect(broadcastArguments[0]).toBe('oTracking.event');
+					expect(broadcastArguments[1]).toMatchSnapshot();
+				});
+
+				describe('when all metrics are sent a second time', () => {
```
Comment (joelcarr) [inline src/client/__test__/realUserMonitoring.spec.js:107]: **todo:** We only need to send any one of the metrics a second time to test the following line: https://github.com/Financial-Times/n-tracking/blob/7e0cffef82d3a7c13552121d86610f84f68118ac/src/client/trackers/realUserMonitoringForPerformance.js#L85

Because of this we should make the test description and setup clear that we are testing if any one of these metrics is sent again, not if all of the metrics are sent again.

```suggestion
				describe('when any one of the metrics are sent a second time', () => {
```

Comment (joelcarr) [reply src/client/__test__/realUserMonitoring.spec.js:107]: **suggestion:** We could also assert that the `realUserMonitoringForPerformance().context[metric]` value has not been updated.

File: src/client/__test__/realUserMonitoring.spec.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+						expect(broadcast).toHaveBeenCalledTimes(1);
+					});
+				});
+			});
+
+		});
+
+		describe('when the seed is not in the sample', () => {
+
+			beforeEach(async () => {
```
Comment (joelcarr) [inline src/client/__test__/realUserMonitoring.spec.js:128]: **question:** Is this async needed? I can't see an `await` in the function body.

**Review by joelcarr — CHANGES_REQUESTED** [review]
Thank you for adding tests @rowanmanning! This is great prep work for your ticket and I think an example we should share with others for "how to refactor something you currently don't understand".

File: src/client/__test__/realUserMonitoring.spec.js (hunk 2 above)
Comment (rowanmanning) [reply src/client/__test__/realUserMonitoring.spec.js]: Nope! It's a copy-paste error 🙂 removing

File: src/client/__test__/realUserMonitoring.spec.js (hunk 1 above)
Comment (rowanmanning) [reply src/client/__test__/realUserMonitoring.spec.js:107]: Agreed on only testing one metric a second time and updating the text (both done). I think because the `context` object is never directly exposed/returned by `realUserMonitoringForPerformance` we probably don't need to test it?

The only time that the properties of `context` are used is when the o-tracking event is broadcast, and this test verifies that the event can only be broadcast once so it's kind of irrelevant to the outside world if the `context` values change

**Comment by rowanmanning** [comment]
Thanks for the thorough review, @joelcarr! 🙂 some well spotted typos and improvements in there. I've made all of the requested changes except one (see comment)

File: src/client/__test__/realUserMonitoring.spec.js (hunk 1 above)
Comment (joelcarr) [reply src/client/__test__/realUserMonitoring.spec.js]: Ahh yes `context` isn't returned so there is no way to assert it. I now understand that as long as the broadcast is not called again, we don't mind if the context somehow changed as it will have no effect on anything else.

**Review by joelcarr — APPROVED** [review]
Captain Jack Sparrow passing over some rum
